### ISSUE TITLE
ENH: implement handling of `epochs` param

### ISSUE THREAD

**Issue by adriangb** [adriangb/scikeras#51]
We need to discuss:

- Naming of the parameter. Keras uses `epochs`, sklearn uses `max_iter`.
- Default value of the parameter. Keras uses `epochs=1`, skelarn classifiers vary:
  - [SGDClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html): 1000
  - [MLPClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html): 200
  - [SVC](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html): -1 (no limit, tolerance based stopping only).
- Use of `initial_epoch`. Keras has no centralized tracking of epochs. It seems that `initial_epoch` can be used for stateless optimizers that depend on the epoch to adjust internal parameters like the learning rate ([SO explanation](https://stackoverflow.com/a/52478034/6582418)). We could enable the parameter, or hide it from the user since we do keep track of the epochs as the length of the values in `BaseWrapper.history_` (i.e. `self.model_.fit(..., initial_epoch=len(self.history_.loss))`)

**Comment by stsievert** [comment]
Here are my opinions:

* Naming: `epochs` because Keras users are more familiar with it, and it's used in the Scikit-learn documentation.*
* Default value: I'm inclined to mirror MLPClassifier because it's also a neural net (not a linear model like SGDClassifier).
* I'm not seeing any reason to include `initial_epoch` because SciKeras could hide it from the user with your suggestion.

<sup>*[SGDRegressor] docs: "The maximum number of passes over the training data (aka epochs)."</sup>

[SGDRegressor]:https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDRegressor.html

**Comment by adriangb** [comment]
Excellent. I'm generally on board with all of these.

Do you think there might be any issues with Scikeras determining `initial_epochs`? That is, could that lead to unexpected behavior? Will we let users override it in some way?

**Comment by stsievert** [comment]
> could that lead to unexpected behavior?

Using `len(self.model_.history["loss"])` seems like a reasonable approximation to me. That might lead to challenges if you're calling `partial_fit` with different sized data bunches, but that could be worked around:

``` python
class BaseWrapper:
    @property
    def initial_epochs(self):
        return len(self.model_.history["loss"])
```

That'd allow users to override that function and do whatever behavior they want.

**Comment by stsievert** [comment]
And SciKeras absolutely needs to track this; without this, `partial_fit` is broken because it always assumes training begins at the first epoch. This means step size decay schedules won't work properly (and step size decay schedules are critical for high performance).

**Comment by adriangb** [comment]
> That'd allow users to override that function and do whatever behavior they want.

I like this. Any user running into problems is probably using a custom optimizer. They should be advanced enough to override this method.

> SciKeras absolutely needs to track initial_epoch; without that, partial_fit is broken

Yes and no. I believe that all optimizers built into Keras keep track of epochs internally, but in general yes, things will be broken without this.

**Comment by adriangb** [comment]
I'd like to amend the above:

* Calling `fit` forces `initial_epoch=0` and `epochs=self.epcohs`, **even if `warm_start=True`**. *
* Calling `partial_fit` forces `epochs=1` and `initial_epoch=len(self.model_.history["loss"])`.

\* `fit` with `warm_start=False` not only resets the `inital_epoch`, it also rebuilds the model. I believe this reflects what sklearn does.

**Comment by stsievert** [comment]
> * Calling fit forces initial_epoch=0 and epochs=self.epcohs, even if warm_start=True
> * Calling partial_fit forces epochs=1 and initial_epoch=len(self.model_.history["loss"]).

Is this a correct rewording?

1. Calling `fit` should train for `self.epochs`. If `warm_start=True` it should reuse the model weights at initialization but restart the optimizer.
2. Calling `partial_fit` should train for 1 epoch.

I'm inclined to make `fit` repeated `partial_fit` calls; I think that's the simplest to explain and what I would expect as a user.

**Comment by adriangb** [comment]
That does sound correct to me.

I think we should continue to let Keras handle the training loop. I imagine repeatedly calling `partial_fit` will get complicated with Keras callbacks, etc. I think we can manage it just like we're doing now but also using the `epochs` and `intial_epoch` parameters.

**Comment by stsievert** [comment]
> I imagine repeatedly calling partial_fit will get complicated with Keras callbacks, etc.

How would having the `fit` function be repeated calls to `partial_fit` get complicated with Keras callbacks?

**Comment by adriangb** [comment]
I'm not 100% sure, just a knee jerk reaction. I would want to really think it through, and maybe check with the TF team about performance, before implementing that.

**Comment by adriangb** [comment]
For now I implemented this using the existing logic (i.e. not making fit repeated partial_fit calls). I'm going to close this issue until a good reason is brought up for changing it (at which point I'll be happy to implement something else).
